(window.webpackJsonp=window.webpackJsonp||[]).push([[4],{191:function(e,t,r){"use strict";r.r(t);var a=r(0),n=Object(a.a)({},function(){var e=this,t=e.$createElement,a=e._self._c||t;return a("ContentSlotsDistributor",{attrs:{"slot-key":e.$parent.slotKey}},[a("h1",{attrs:{id:"bioinformatic-overview"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#bioinformatic-overview","aria-hidden":"true"}},[e._v("#")]),e._v(" Bioinformatic Overview")]),e._v(" "),a("p",[e._v("The three main functions of the pipeline are:")]),e._v(" "),a("ul",[a("li",[e._v("Sequencing read alignment")]),e._v(" "),a("li",[e._v("Somatic variant detection")]),e._v(" "),a("li",[e._v("Germline variant detection")])]),e._v(" "),a("p",[e._v("Below are described the separate Nextflow processes and associated with each module and the tools used. The following diagram outlines the workflow:")]),e._v(" "),a("img",{attrs:{id:"diagram",src:r(69)}}),e._v(" "),a("p",[a("small",[e._v("Note: The pipeline can be run with already-aligned BAM files as input, which avoids the first of these three modules.")])]),e._v(" "),a("h2",{attrs:{id:"read-alignment"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#read-alignment","aria-hidden":"true"}},[e._v("#")]),e._v(" Read alignment")]),e._v(" "),a("p",[e._v("Vaporware accepts as input sequencing reads from one or multiple FASTQ file pairs (corresponding to separate sequencing lanes) per sample, as "),a("router-link",{attrs:{to:"/run-pipeline.html#the-mapping-file"}},[e._v("described")]),e._v(". These are aligned against the human genome using common practices.")],1),e._v(" "),a("ul",[a("li",[a("strong",[e._v("AlignReads")]),e._v(": Alignment using "),a("a",{attrs:{href:"http://bio-bwa.sourceforge.net/",target:"_blank",rel:"noopener noreferrer"}},[e._v("BWA mem"),a("OutboundLink")],1),e._v(", followed by conversion to BAM file format and sorting using "),a("a",{attrs:{href:"https://samtools.github.io",target:"_blank",rel:"noopener noreferrer"}},[e._v("samtools"),a("OutboundLink")],1),e._v(". Additionally, this generates FASTQ QC metrics using "),a("a",{attrs:{href:"https://github.com/OpenGene/fastp",target:"_blank",rel:"noopener noreferrer"}},[e._v("fastp"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("li",[a("strong",[e._v("MergeBams")]),e._v(": Merging of BAM files across sequencing lanes using "),a("a",{attrs:{href:"https://samtools.github.io",target:"_blank",rel:"noopener noreferrer"}},[e._v("samtools"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("li",[a("strong",[e._v("MarkDuplicates")]),e._v(": PCR-duplicate marking using "),a("a",{attrs:{href:"https://software.broadinstitute.org/gatk",target:"_blank",rel:"noopener noreferrer"}},[e._v("GATK MarkDuplicates"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("li",[a("strong",[e._v("CreateRecalibrationTable")]),e._v(" and "),a("strong",[e._v("RecalibrateBam")]),e._v(": Base-quality score recalibration with "),a("a",{attrs:{href:"https://software.broadinstitute.org/gatk/",target:"_blank",rel:"noopener noreferrer"}},[e._v("GATK BaseRecalibrator and ApplyBQSR"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("li",[a("strong",[e._v("Alfred")]),e._v(": BAM file QC using "),a("a",{attrs:{href:"https://github.com/tobiasrausch/alfred",target:"_blank",rel:"noopener noreferrer"}},[e._v("Alfred"),a("OutboundLink")],1),e._v(".")])]),e._v(" "),a("h2",{attrs:{id:"somatic-analyses"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#somatic-analyses","aria-hidden":"true"}},[e._v("#")]),e._v(" Somatic analyses")]),e._v(" "),a("ul",[a("li",[e._v("SNVs and InDels are called using "),a("a",{attrs:{href:"https://software.broadinstitute.org/gatk/documentation/tooldocs/4.beta.4/org_broadinstitute_hellbender_tools_walkers_mutect_Mutect2.php",target:"_blank",rel:"noopener noreferrer"}},[e._v("MuTect2"),a("OutboundLink")],1),e._v(" and "),a("a",{attrs:{href:"https://github.com/Illumina/strelka",target:"_blank",rel:"noopener noreferrer"}},[e._v("Strelka2"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("li",[e._v("Structural variants are detected by "),a("a",{attrs:{href:"https://github.com/dellytools/delly",target:"_blank",rel:"noopener noreferrer"}},[e._v("Delly"),a("OutboundLink")],1),e._v(" and "),a("a",{attrs:{href:"https://github.com/Illumina/manta",target:"_blank",rel:"noopener noreferrer"}},[e._v("Manta"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("li",[e._v("Copy-number analysis is performed with "),a("a",{attrs:{href:"https://github.com/mskcc/facets",target:"_blank",rel:"noopener noreferrer"}},[e._v("FACETS"),a("OutboundLink")],1),e._v(" and processed using "),a("a",{attrs:{href:"https://github.com/mskcc/facets-suite",target:"_blank",rel:"noopener noreferrer"}},[e._v("facets-suite"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("li",[e._v("Microsatellite instability is detected using "),a("a",{attrs:{href:"https://github.com/ding-lab/msisensor",target:"_blank",rel:"noopener noreferrer"}},[e._v("MSIsensor"),a("OutboundLink")],1),e._v(".")])]),e._v(" "),a("h2",{attrs:{id:"germline-analyses"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#germline-analyses","aria-hidden":"true"}},[e._v("#")]),e._v(" Germline analyses")]),e._v(" "),a("ul",[a("li",[e._v("SNVs and InDels are called using "),a("a",{attrs:{href:"https://software.broadinstitute.org/gatk/documentation/tooldocs/4.beta.4/org_broadinstitute_hellbender_tools_walkers_mutect_Mutect2.php",target:"_blank",rel:"noopener noreferrer"}},[e._v("MuTect2"),a("OutboundLink")],1),e._v(" and "),a("a",{attrs:{href:"https://github.com/Illumina/strelka",target:"_blank",rel:"noopener noreferrer"}},[e._v("Strelka2"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("li",[e._v("Structural variants are detected by "),a("a",{attrs:{href:"https://github.com/dellytools/delly",target:"_blank",rel:"noopener noreferrer"}},[e._v("Delly"),a("OutboundLink")],1),e._v(" and "),a("a",{attrs:{href:"https://github.com/Illumina/manta",target:"_blank",rel:"noopener noreferrer"}},[e._v("Manta"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("li",[e._v("We currently do not have a special module dedicated to germline CNVs besides that provided by the germline SV calls.")])]),e._v(" "),a("h2",{attrs:{id:"wgs-vs-wes"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#wgs-vs-wes","aria-hidden":"true"}},[e._v("#")]),e._v(" WGS vs. WES")]),e._v(" "),a("p",[e._v("Contact us if you are interest in support for other sequencing assays or capture kits.")])])},[],!1,null,null,null);t.default=n.exports},69:function(e,t,r){e.exports=r.p+"assets/img/pipeline-flowchart.dd725071.png"}}]);