<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>gnomAD</title>
    <meta name="description" content="">
    
    
    <link rel="preload" href="/assets/css/0.styles.d4613532.css" as="style"><link rel="preload" href="/assets/js/app.98741281.js" as="script"><link rel="preload" href="/assets/js/2.e10afc37.js" as="script"><link rel="preload" href="/assets/js/10.47a85e35.js" as="script"><link rel="prefetch" href="/assets/js/11.e78ae733.js"><link rel="prefetch" href="/assets/js/12.569b3204.js"><link rel="prefetch" href="/assets/js/13.4943b92b.js"><link rel="prefetch" href="/assets/js/14.c1aef2a4.js"><link rel="prefetch" href="/assets/js/15.88b928ad.js"><link rel="prefetch" href="/assets/js/16.8466b006.js"><link rel="prefetch" href="/assets/js/17.48c6fe4c.js"><link rel="prefetch" href="/assets/js/18.38a33712.js"><link rel="prefetch" href="/assets/js/19.b6fe101c.js"><link rel="prefetch" href="/assets/js/20.63f9a4f8.js"><link rel="prefetch" href="/assets/js/21.05021021.js"><link rel="prefetch" href="/assets/js/3.cd2560e0.js"><link rel="prefetch" href="/assets/js/4.1367f42d.js"><link rel="prefetch" href="/assets/js/5.70f8d865.js"><link rel="prefetch" href="/assets/js/6.a2a4a4fd.js"><link rel="prefetch" href="/assets/js/7.22b6c02c.js"><link rel="prefetch" href="/assets/js/8.d5159c77.js"><link rel="prefetch" href="/assets/js/9.40da1173.js">
    <link rel="stylesheet" href="/assets/css/0.styles.d4613532.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container no-sidebar"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/" class="home-link router-link-active"></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <!----></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><!---->  <!----> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="gnomad"><a href="#gnomad" aria-hidden="true" class="header-anchor">#</a> gnomAD</h1> <p><small>As of 2019-08-16, the latest release of gnomAD is 2.1.1 from March 6, 2019.</small></p> <p>Variant-level allele counts and frequencies from gnomAD's exome and genome cohorts are used to annotate somatic and germline SNVs/indels.</p> <p>These files were retrieved and processed as such:</p> <h2 id="exomes"><a href="#exomes" aria-hidden="true" class="header-anchor">#</a> Exomes</h2> <p>The exome VCF file contains a lot of information. We prune most of this to reduce file size and only keep relevant information. Only values from the non-cancer subset of the total population are used, since this excludes the normals from TCGA.</p> <div class="language-shell extra-class"><pre class="language-text"><code># Download files
wget https://storage.googleapis.com/gnomad-public/release/2.1.1/vcf/exomes/gnomad.exomes.r2.1.1.sites.vcf.bgz
wget https://storage.googleapis.com/gnomad-public/release/2.1.1/vcf/exomes/gnomad.exomes.r2.1.1.sites.vcf.bgz.tbi
mv gnomad.exomes.r2.1.1.sites.vcf.bgz gnomad.exomes.r2.1.1.sites.vcf.gz
mv gnomad.exomes.r2.1.1.sites.vcf.bgz.tbi gnomad.exomes.r2.1.1.sites.vcf.gz.tbi

# Parse INFO columns to retain only relevant ones
bcftools view --header-only gnomad.exomes.r2.1.1.sites.vcf.gz | \
    grep -E &quot;non_cancer_AC|non_cancer_AF&quot; | \
    grep -v -e &quot;_male&quot; -e &quot;_female&quot; \
    &gt; gnomad.exomes.r2.1.1.sites.retained.info

paste &lt;(grep -oP &quot;(?&lt;=ID\=)[A-Za-z_]+&quot; gnomad.exomes.r2.1.1.sites.retained.info) \
      &lt;(grep -oP &quot;(?&lt;=Description=\&quot;)[A-Za-z\(\),\-\_\ ]+&quot; gnomad.exomes.r2.1.1.sites.retained.info) \
      &gt; tmp &amp;&amp; \
      mv tmp gnomad.exomes.r2.1.1.sites.retained.info

COLS=$(Rscript -e &quot;out = paste0('^INFO/', paste(read.delim('gnomad.exomes.r2.1.1.sites.retained.info', header = F)[['V1']], collapse = ',INFO/')); cat(out)&quot;)

# Apply this, and retain filtered sites
bcftools annotate \
    --remove &quot;$COLS&quot; \
    --include 'FILTER~&quot;PASS&quot; | FILTER~&quot;RF&quot;' \
    --output-type z \
    --output tmp.vcf.gz \
    gnomad.exomes.r2.1.1.sites.vcf.gz

tabix --preset vcf tmp.vcf.gz

# Mark filtered sites
bcftools annotate \
    --annotations gnomad.exomes.r2.1.1.sites.non_cancer.vcf.gz \
    --include 'FILTER!=&quot;PASS&quot;' \
    --mark-sites &quot;+gnomAD_FILTER&quot; \
    -k \
    --output-type z \
    --output gnomad.exomes.r2.1.1.sites.non_cancer.vcf.gz \
    tmp.vcf.gz

tabix --preset vcf gnomad.exomes.r2.1.1.sites.non_cancer.vcf.gz

# Clean up
rm gnomad.exomes.r2.1.1.sites.retained.info
rm tmp.vcf.gz tmp.vcf.gz.tbi
</code></pre></div><h2 id="genomes"><a href="#genomes" aria-hidden="true" class="header-anchor">#</a> Genomes</h2> <p>For genomes, there is no non-cancer subset.</p> <div class="language-shell extra-class"><pre class="language-text"><code># Download, one chromosome at the time
for chr in {1..22} X
do
    wget https://storage.googleapis.com/gnomad-public/release/2.1.1/vcf/genomes/gnomad.genomes.r2.1.1.sites.${chr}.vcf.bgz
    wget https://storage.googleapis.com/gnomad-public/release/2.1.1/vcf/genomes/gnomad.genomes.r2.1.1.sites.${chr}.vcf.bgz.tbi
    mv gnomad.genomes.r2.1.1.sites.${chr}.vcf.bgz gnomad.genomes.r2.1.1.sites.${chr}.vcf.gz
    mv gnomad.genomes.r2.1.1.sites.${chr}.vcf.bgz.tbi gnomad.genomes.r2.1.1.sites.${chr}.vcf.gz.tbi
done

# Parse INFO columns to retain only relevant ones
bcftools view --header-only gnomad.genomes.r2.1.1.sites.1.vcf.gz | \
    grep -E &quot;AC|AF&quot; | \
    grep -v -e &quot;_male&quot; -e &quot;_female&quot; -e &quot;controls&quot; -e &quot;topmed&quot; -e &quot;neuro&quot; -e &quot;vep&quot; -e &quot;raw&quot; -e &quot;=popmax&quot; -e &quot;AC0&quot; | \
    cut -f3 -d&quot;=&quot; | cut -f1 -d&quot;,&quot; \
    &gt; gnomad.genomes.r2.1.1.sites.retained.info

COLS=$(Rscript -e &quot;out = paste0('^INFO/', paste(read.delim('gnomad.genomes.r2.1.1.sites.retained.info', header = F)[['V1']], collapse = ',INFO/')); cat(out)&quot;)
CHR=({1..22} X)

# Apply this, and retain filtered sites
for chr in ${CHR[@]}
do
    bcftools annotate \
    --remove &quot;$COLS&quot; \
    --include 'FILTER~&quot;PASS&quot; | FILTER~&quot;RF&quot;' \
    --output-type z \
    --output tmp.${chr}.vcf.gz \
    gnomad.genomes.r2.1.1.sites.${chr}.vcf.gz

    tabix --preset vcf tmp.${chr}.vcf.gz

    bcftools annotate \
    --annotations tmp.${chr}.vcf.gz \
    --include 'FILTER!=&quot;PASS&quot;' \
    --mark-sites &quot;+gnomAD_FILTER&quot; \
    -k \
    --output-type z \
    --output gnomad.genomes.r2.1.1.sites.${chr}.minimal.vcf.gz \
    tmp.${chr}.vcf.gz

    tabix --preset vcf gnomad.genomes.r2.1.1.sites.${chr}.minimal.vcf.gz
done

# Concatenate to one file
bcftools concat \
    --output-type z \
    --output gnomad.genomes.r2.1.1.sites.minimal.vcf.gz \
    gnomad.genomes.r2.1.1.sites.{1..22}.minimal.vcf.gz \
    gnomad.genomes.r2.1.1.sites.X.minimal.vcf.gz

tabix --preset vcf gnomad.genomes.r2.1.1.sites.minimal.vcf.gz

# Clean up
for chr in ${CHR[@]}
do
    rm gnomad.genomes.r2.1.1.sites.${chr}.minimal.vcf.gz*
    rm tmp.${chr}.vcf.gz*
done
</code></pre></div></div> <footer class="page-edit"><!----> <!----></footer> <!----> </main></div><div class="global-ui"></div></div>
    <script src="/assets/js/app.98741281.js" defer></script><script src="/assets/js/2.e10afc37.js" defer></script><script src="/assets/js/10.47a85e35.js" defer></script>
  </body>
</html>
