<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Outputs</title>
    <meta name="description" content="">
    
    
    <link rel="preload" href="/assets/css/0.styles.d4613532.css" as="style"><link rel="preload" href="/assets/js/app.98741281.js" as="script"><link rel="preload" href="/assets/js/2.e10afc37.js" as="script"><link rel="preload" href="/assets/js/14.c1aef2a4.js" as="script"><link rel="prefetch" href="/assets/js/10.47a85e35.js"><link rel="prefetch" href="/assets/js/11.e78ae733.js"><link rel="prefetch" href="/assets/js/12.569b3204.js"><link rel="prefetch" href="/assets/js/13.4943b92b.js"><link rel="prefetch" href="/assets/js/15.88b928ad.js"><link rel="prefetch" href="/assets/js/16.8466b006.js"><link rel="prefetch" href="/assets/js/17.48c6fe4c.js"><link rel="prefetch" href="/assets/js/18.38a33712.js"><link rel="prefetch" href="/assets/js/19.b6fe101c.js"><link rel="prefetch" href="/assets/js/20.63f9a4f8.js"><link rel="prefetch" href="/assets/js/21.05021021.js"><link rel="prefetch" href="/assets/js/3.cd2560e0.js"><link rel="prefetch" href="/assets/js/4.1367f42d.js"><link rel="prefetch" href="/assets/js/5.70f8d865.js"><link rel="prefetch" href="/assets/js/6.a2a4a4fd.js"><link rel="prefetch" href="/assets/js/7.22b6c02c.js"><link rel="prefetch" href="/assets/js/8.d5159c77.js"><link rel="prefetch" href="/assets/js/9.40da1173.js">
    <link rel="stylesheet" href="/assets/css/0.styles.d4613532.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container no-sidebar"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/" class="home-link router-link-active"></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <!----></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><!---->  <!----> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="outputs"><a href="#outputs" aria-hidden="true" class="header-anchor">#</a> Outputs</h1> <p>All paths below are relative to the base directory <code>outDir</code> as described in the <a href="/running-the-pipeline.html">run instructions</a>.</p> <div class="language-shell extra-class"><pre class="language-text"><code>outDir
├── bams
├── qc
├── somatic
└── germline
</code></pre></div><h2 id="bam-files"><a href="#bam-files" aria-hidden="true" class="header-anchor">#</a> BAM Files</h2> <p>The <code>bams</code> folder contains the final aligned and post-processed BAM files along with index files.</p> <h2 id="qc-outputs"><a href="#qc-outputs" aria-hidden="true" class="header-anchor">#</a> QC Outputs</h2> <p>FASTQ file, read alignment and basic BAM file QC is in the <code>qc</code> directory:</p> <div class="language-shell extra-class"><pre class="language-text"><code>qc
├── alfred
├── collecthsmetrics
├── conpair
├── fastp
└── alignment_qc.txt
</code></pre></div><p>These outputs are:</p> <ul><li><code>fastp</code> (folder): An HTML report for each FASTQ file pair per sample.</li> <li><code>alfred</code> (folder): A per-sample and per-readgroup BAM file alignment metrics in text and PDF files.</li> <li><code>collectshsmetrics</code> (folder): For exomes, per-sample hybridisation-selection metrics in the.</li> <li><code>conpair</code> (folder): Per tumor-normal-pair contamination and sample concordance estimates.</li> <li><code>alignment_qc.txt</code>: Aggregated read-alignments statistics file, from the <code>alfred</code> and <code>collectshsmetrics</code> folders.</li></ul> <h2 id="somatic-data"><a href="#somatic-data" aria-hidden="true" class="header-anchor">#</a> Somatic data</h2> <p>The result of the somatic analyses is output in summarized forms in the <code>somatic</code> folder:</p> <div class="language-shell extra-class"><pre class="language-text"><code>somatic
├── cna
├── mut_somatic.maf
├── mut_somatic_neoantigens.txt
├── cna_armlevel.txt
├── cna_genelevel.txt
├── cna_hisens.seg
├── cna_purity.seg
├── cna_facets_run_info.txt
├── sv_somatic.vcf.gz
└── merged_metadata.txt
</code></pre></div><p>These outputs are:</p> <ul><li><code>facets</code> (folder): Individual copy-number profiles from FACETS, per tumor-normal pair.</li> <li><code>mut_somatic.maf</code>: Filtered mutations from MuTect2 and Strelka2, annotated with mutational effects, neoantigen predictions, and zygosity, as <a href="/variant-annotation-and-filtering.html#somatic-snvs-and-indels">described elsewhere</a>.</li> <li><code>mut_somatic_neoantigens.txt</code>: Neoantigen predictions from NetMHCpan for all samples.</li> <li><code>cna_armlevel.txt</code>, <code>cna_genelevel.txt</code>, and <code>cna_hisens.seg</code>, <code>cna_purity.seg</code>, and <code>cna_facets_run_info.txt</code>, summarized arm- and gene-level output from Facets, as well as IGV-style segmentation files and Facets run information.</li> <li><code>sv_somatic.vcf.gz</code>: All structural variants detected by Delly and Manta.</li> <li><code>merged_metadata.txt</code>: Merged metadata across samples and analyses.</li></ul> <h2 id="germline-data"><a href="#germline-data" aria-hidden="true" class="header-anchor">#</a> Germline data</h2> <p>The result of the germline analyses is output in summarized forms in the <code>germline</code> folder:</p> <div class="language-shell extra-class"><pre class="language-text"><code>germline/
├── mut_germline.maf
└── sv_germline.vcf.gz
</code></pre></div><p>These outputs are:</p> <ul><li><code>mut_germline.maf</code>: Filtered mutations from HaplotypeCaller and Strelka2, annotated with mutational effects and zygosity, as <a href="/variant-annotation-and-filtering.html#germline-snvs-and-indels">described elsewhere</a>.</li> <li><code>sv_germline.vcf.gz</code>: All structural variants Delly and Manta.</li></ul> <h2 id="extended-outputs"><a href="#extended-outputs" aria-hidden="true" class="header-anchor">#</a> Extended Outputs</h2> <h1 id="under-development"><a href="#under-development" aria-hidden="true" class="header-anchor">#</a> <em>Under development.</em></h1> <p>When run with the flag <code>--publishAll</code>, the pipeline will output additional intermediate data from select processes. These are:</p> <div class="language-shell extra-class"><pre class="language-text"><code>somatic
├── mutations
    └── mutect2
    └── strelka2
├── structural_variants
    └── delly
    └── manta    
├── facets
└── lohhla

germline
├── mutations
    └── haplotypecaller
    └── strelka2
└── structural_variants
    └── delly
    └── manta
</code></pre></div><p>The <code>mutations</code> subdirectory contain VCFs with the unfiltered variant calls from the somatic and germline SNV/indel and SV callers. Additionally, these directories contain per-sample unfiltered MAF files generated in the <code>SomaticAnnotateMaf</code> and <code>GermlineAnnotateMaf</code> process, respectively. The <code>facets</code> subdirectory will contain the full arm- and gene-level outputs per sample. In the <code>lohhla</code> subdirectory the full LOHHLA LOH output metrics will be together with a PDF file with graphical output.</p></div> <footer class="page-edit"><!----> <!----></footer> <!----> </main></div><div class="global-ui"></div></div>
    <script src="/assets/js/app.98741281.js" defer></script><script src="/assets/js/2.e10afc37.js" defer></script><script src="/assets/js/14.c1aef2a4.js" defer></script>
  </body>
</html>
