<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Reference Resources</title>
    <meta name="description" content="">
    
    
    <link rel="preload" href="/assets/css/0.styles.d4613532.css" as="style"><link rel="preload" href="/assets/js/app.98741281.js" as="script"><link rel="preload" href="/assets/js/2.e10afc37.js" as="script"><link rel="preload" href="/assets/js/15.88b928ad.js" as="script"><link rel="prefetch" href="/assets/js/10.47a85e35.js"><link rel="prefetch" href="/assets/js/11.e78ae733.js"><link rel="prefetch" href="/assets/js/12.569b3204.js"><link rel="prefetch" href="/assets/js/13.4943b92b.js"><link rel="prefetch" href="/assets/js/14.c1aef2a4.js"><link rel="prefetch" href="/assets/js/16.8466b006.js"><link rel="prefetch" href="/assets/js/17.48c6fe4c.js"><link rel="prefetch" href="/assets/js/18.38a33712.js"><link rel="prefetch" href="/assets/js/19.b6fe101c.js"><link rel="prefetch" href="/assets/js/20.63f9a4f8.js"><link rel="prefetch" href="/assets/js/21.05021021.js"><link rel="prefetch" href="/assets/js/3.cd2560e0.js"><link rel="prefetch" href="/assets/js/4.1367f42d.js"><link rel="prefetch" href="/assets/js/5.70f8d865.js"><link rel="prefetch" href="/assets/js/6.a2a4a4fd.js"><link rel="prefetch" href="/assets/js/7.22b6c02c.js"><link rel="prefetch" href="/assets/js/8.d5159c77.js"><link rel="prefetch" href="/assets/js/9.40da1173.js">
    <link rel="stylesheet" href="/assets/css/0.styles.d4613532.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container no-sidebar"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/" class="home-link router-link-active"></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <!----></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><!---->  <!----> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="reference-resources"><a href="#reference-resources" aria-hidden="true" class="header-anchor">#</a> Reference Resources</h1> <p>This and associated pages in this section provide details on provenance and generation of all reference files used in <code>pipeline.nf</code>. Usage of these files is defined in the <a href="../conf/references.config">references configuration file</a>.</p> <div class="tip custom-block"><p class="custom-block-title">Note</p> <p>All reference files described herein are in assembly GRCh37/hg19 of the human genome.</p></div> <h2 id="genome-assembly"><a href="#genome-assembly" aria-hidden="true" class="header-anchor">#</a> Genome Assembly</h2> <p>Part of the <a href="https://software.broadinstitute.org/gatk/download/bundle" target="_blank" rel="noopener noreferrer">GATK bundle<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>, also available <a href="https://console.cloud.google.com/storage/browser/gatk-legacy-bundles/b37" target="_blank" rel="noopener noreferrer">here<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>. Vaporware uses the <strong>human_g1k_v37_decoy</strong> assembly of the genome.</p> <h2 id="genomic-intervals"><a href="#genomic-intervals" aria-hidden="true" class="header-anchor">#</a> Genomic Intervals</h2> <p>BED files that specify the regions of the genome to consider for variant calling are specified in the <a href="/running-the-pipeline.html#input-files">input files</a>.</p> <h3 id="exome-capture-platforms"><a href="#exome-capture-platforms" aria-hidden="true" class="header-anchor">#</a> Exome Capture Platforms</h3> <p>For exomes, use BED file corresponding to the platform used for target capture. Currently, Vaporware supports:</p> <ul><li><strong>AgilentExon_51MB</strong>: SureSelectXT Human All Exon V4 from Agilent.</li> <li><strong>IDT_Exome</strong>: xGen Exome Research Panel v1.0 from IDT.</li></ul> <div class="tip custom-block"><p class="custom-block-title">Note</p> <p>Contact us if you are interested in support for other sequencing assays or capture kits.</p></div> <p>The bait and target files are provided by the kit manufacturer. These are used to estimate bait- and target-level coverage metrics as well as for variant calling.</p> <p>We add 5 bp to each end of exons in the target file to make sure splice site mutations can be called:</p> <div class="language-shell extra-class"><pre class="language-text"><code>bedtools slop \
    -g b37.chrom.sizes \
    -i targets.bed \
    -r 5 \
    -l 5 \
    &gt; targets.plus5bp.bed
</code></pre></div><h3 id="callable-regions-for-genomes"><a href="#callable-regions-for-genomes" aria-hidden="true" class="header-anchor">#</a> Callable Regions for Genomes</h3> <p>For genomes, a list of &quot;callable&quot; regions from GATK's bundle is used. This is converted from an interval list to a BED file:</p> <div class="language-shell extra-class"><pre class="language-text"><code>gatk IntervalListToBed \
    --INPUT b37_wgs_calling_regions.v1.interval_list \
    --OUTPUT b37_wgs_calling_regions.v1.bed
</code></pre></div><h2 id="repeatmasker-and-mappability-blacklist"><a href="#repeatmasker-and-mappability-blacklist" aria-hidden="true" class="header-anchor">#</a> RepeatMasker and Mappability Blacklist</h2> <p>BED files with genomic repeat and mappability information are used to annotate the VCFs with somatic and germline SNV/indels. These data are from <a href="http://www.repeatmasker.org/" target="_blank" rel="noopener noreferrer">RepeatMasker<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a> and the <a href="http://rohsdb.cmb.usc.edu/GBshape/ENCODE/index.html" target="_blank" rel="noopener noreferrer">ENCODE consortium<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>, and the files are retrieved from the <a href="https://genome.ucsc.edu" target="_blank" rel="noopener noreferrer">UCSC Genome Browser<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a> and parsed as such:</p> <div class="language-shell extra-class"><pre class="language-text"><code>wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/rmsk.txt.gz
gunzip rmsk.txt.gz
cut -f6-8,12 rmsk.txt | \
    grep -e &quot;Low_complexity&quot; -e &quot;Simple_repeat&quot; | \
    sed 's/^chr//g'&gt; rmsk_mod.bed
bgzip rmsk_mod.bed
tabix --preset bed rmsk_mod.bed.gz

wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/wgEncodeDacMapabilityConsensusExcludable.bed.gz
gunzip wgEncodeDacMapabilityConsensusExcludable.bed.gz
sed -i 's/^chr//g' wgEncodeDacMapabilityConsensusExcludable.bed
bgzip wgEncodeDacMapabilityConsensusExcludable.bed
tabix --preset bed wgEncodeDacMapabilityConsensusExcludable.bed.gz
</code></pre></div><h2 id="preferred-transcript-isoforms"><a href="#preferred-transcript-isoforms" aria-hidden="true" class="header-anchor">#</a> Preferred Transcript Isoforms</h2> <p>The <code>--custom-enst</code> argument to vcf2maf takes a list of preferred gene transcript isoforms which mutations are mapped onto. We supply a consensus list of <a href="https://github.com/mskcc/vcf2maf/tree/master/data" target="_blank" rel="noopener noreferrer"><code>isoform_overrides_at_mskcc</code> and <code>isoform_overrides_uniprot</code><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>, generated as such:</p> <div class="language-r extra-class"><pre class="language-r"><code>t1 <span class="token operator">=</span> readr<span class="token operator">::</span>read_tsv<span class="token punctuation">(</span><span class="token string">'isoform_overrides_at_mskcc'</span><span class="token punctuation">)</span>
t2 <span class="token operator">=</span> readr<span class="token operator">::</span>read_tsv<span class="token punctuation">(</span><span class="token string">'isoform_overrides_uniprot'</span><span class="token punctuation">)</span>
t2 <span class="token percent-operator operator">%&gt;%</span>
    dplyr<span class="token operator">::</span>filter<span class="token punctuation">(</span>gene_name <span class="token percent-operator operator">%nin%</span> t1<span class="token operator">$</span>gene_name<span class="token punctuation">)</span> <span class="token percent-operator operator">%&gt;%</span>
    dplyr<span class="token operator">::</span>bind_rows<span class="token punctuation">(</span>.<span class="token punctuation">,</span> t1<span class="token punctuation">)</span> <span class="token percent-operator operator">%&gt;%</span>
    readr<span class="token operator">::</span>write_tsv<span class="token punctuation">(</span><span class="token string">'isoforms'</span><span class="token punctuation">)</span>
</code></pre></div><h2 id="hotspot-annotation"><a href="#hotspot-annotation" aria-hidden="true" class="header-anchor">#</a> Hotspot Annotation</h2> <p>Three types of mutation hotspots are annotated in the somatic MAF. These include SNV, indel in linear space as well as SNV hotspots in 3D space. These are annotated with the <a href="https://github.com/taylor-lab/annotateMaf" target="_blank" rel="noopener noreferrer">annotateMaf package<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>.</p> <h2 id="oncokb-annotation"><a href="#oncokb-annotation" aria-hidden="true" class="header-anchor">#</a> OncoKB Annotation</h2> <p>Functional mutation effects and predicted oncogenicity of variants, as well as level of clinical actionability are from <a href="https://oncokb.org" target="_blank" rel="noopener noreferrer">OncoKB<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a> and annotated using the <a href="https://github.com/oncokb/oncokb-annotator" target="_blank" rel="noopener noreferrer">OncoKB annotator<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>.</p> <h2 id="brca-exchange-annotation"><a href="#brca-exchange-annotation" aria-hidden="true" class="header-anchor">#</a> BRCA Exchange Annotation</h2> <p>Annotation of germline variants in <em>BRCA1</em> and <em>BRCA2</em> is carried out with the <a href="https://github.com/taylor-lab/annotateMaf" target="_blank" rel="noopener noreferrer">annotateMaf package<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>. This includes variant-level annotation from the ENIGMA consortium and ClinVar.</p> <h2 id="structural-variant-calling"><a href="#structural-variant-calling" aria-hidden="true" class="header-anchor">#</a> Structural Variant Calling</h2> <p>Delly provides and takes as an argument a <a href="https://github.com/dellytools/delly/tree/master/excludeTemplates" target="_blank" rel="noopener noreferrer">file of regions<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a> to <em>exclude</em> from variant calling. This excludes telomeres and centromeres from auto- and allosomes as well as any other contig.</p> <p>For Manta, subtract these regions from a bed file of the whole genome to generate a list of regions to <em>include</em>. First clean up the file provided by Delly, since it is not in <code>bed</code> format:</p> <div class="language-shell extra-class"><pre class="language-text"><code>grep -Ev &quot;chr|MT|GL00|NC|hs37d5&quot; human.hg19.excl.tsv &gt; human.hg19.excl.clean.bed
bedtools subtract -a b37.bed -b human.hg19.excl.clean.bed &gt; b37.minusDellyExclude.bed
</code></pre></div></div> <footer class="page-edit"><!----> <!----></footer> <!----> </main></div><div class="global-ui"></div></div>
    <script src="/assets/js/app.98741281.js" defer></script><script src="/assets/js/2.e10afc37.js" defer></script><script src="/assets/js/15.88b928ad.js" defer></script>
  </body>
</html>
