<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Variant Annotation and Filtering | Vaporware</title>
    <meta name="description" content="Computational Pipeline for Whole-Genome and Whole-Exome Sequencing">
    
    
    <link rel="preload" href="/assets/css/0.styles.d4613532.css" as="style"><link rel="preload" href="/assets/js/app.2b579abf.js" as="script"><link rel="preload" href="/assets/js/2.e10afc37.js" as="script"><link rel="preload" href="/assets/js/18.fedd8970.js" as="script"><link rel="prefetch" href="/assets/js/10.01ffc8cf.js"><link rel="prefetch" href="/assets/js/11.e78ae733.js"><link rel="prefetch" href="/assets/js/12.77745fbe.js"><link rel="prefetch" href="/assets/js/13.4943b92b.js"><link rel="prefetch" href="/assets/js/14.04cc32f4.js"><link rel="prefetch" href="/assets/js/15.dc8f5479.js"><link rel="prefetch" href="/assets/js/16.839c34d9.js"><link rel="prefetch" href="/assets/js/17.5b7bd133.js"><link rel="prefetch" href="/assets/js/19.81834ef1.js"><link rel="prefetch" href="/assets/js/20.63f9a4f8.js"><link rel="prefetch" href="/assets/js/21.05021021.js"><link rel="prefetch" href="/assets/js/3.cd2560e0.js"><link rel="prefetch" href="/assets/js/4.253b9be8.js"><link rel="prefetch" href="/assets/js/5.70f8d865.js"><link rel="prefetch" href="/assets/js/6.f8c136e9.js"><link rel="prefetch" href="/assets/js/7.22b6c02c.js"><link rel="prefetch" href="/assets/js/8.f17d8fa6.js"><link rel="prefetch" href="/assets/js/9.f50389f8.js">
    <link rel="stylesheet" href="/assets/css/0.styles.d4613532.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/" class="home-link router-link-active"><img src="/vaporware_stairs.jpg" alt="Vaporware" class="logo"> <span class="site-name can-hide">Vaporware</span></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <nav class="nav-links can-hide"> <a href="https://github.com/mskcc/vaporware" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"> <a href="https://github.com/mskcc/vaporware" target="_blank" rel="noopener noreferrer" class="repo-link">
    GitHub
    <svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a></nav>  <ul class="sidebar-links"><li><a href="/" class="sidebar-link">Home</a></li><li><section class="sidebar-group depth-0"><p class="sidebar-heading"><span>Setup</span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/installation.html" class="sidebar-link">Installation</a></li><li><a href="/juno-setup.html" class="sidebar-link">Juno Setup</a></li><li><a href="/aws-setup.html" class="sidebar-link">AWS Setup</a></li></ul></section></li><li><section class="sidebar-group depth-0"><p class="sidebar-heading"><span>Usage</span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/running-the-pipeline.html" class="sidebar-link">Running the Pipeline</a></li><li><a href="/nextflow-basics.html" class="sidebar-link">Nextflow Basics</a></li><li><a href="/working-with-containers.html" class="sidebar-link">Working With Containers</a></li></ul></section></li><li><a href="/outputs.html" class="sidebar-link">Outputs</a></li><li><a href="/bioinformatics-components.html" class="sidebar-link">Bioinformatics Components</a></li><li><a href="/reference-resources.html" class="sidebar-link">Reference resources</a></li><li><a href="/variant-annotation-and-filtering.html" class="active sidebar-link">Variant Annotation and Filtering</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/variant-annotation-and-filtering.html#somatic-snvs-and-indels" class="sidebar-link">Somatic SNVs and Indels</a></li><li class="sidebar-sub-header"><a href="/variant-annotation-and-filtering.html#germline-snvs-and-indels" class="sidebar-link">Germline SNVs and Indels</a></li><li class="sidebar-sub-header"><a href="/variant-annotation-and-filtering.html#somatic-and-germline-svs" class="sidebar-link">Somatic and Germline SVs</a></li></ul></li><li><section class="sidebar-group depth-0"><p class="sidebar-heading"><span>Help and Other Resources</span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/troubleshooting.html" class="sidebar-link">Troubleshooting</a></li><li><a href="/aws-glossary.html" class="sidebar-link">AWS Glossary</a></li><li><a href="/contributing-to-vaporware.html" class="sidebar-link">Contributing to Vaporware</a></li></ul></section></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="variant-annotation-and-filtering"><a href="#variant-annotation-and-filtering" aria-hidden="true" class="header-anchor">#</a> Variant Annotation and Filtering</h1> <h2 id="somatic-snvs-and-indels"><a href="#somatic-snvs-and-indels" aria-hidden="true" class="header-anchor">#</a> Somatic SNVs and Indels</h2> <p>Variant-level annotation, filtering, and flagging of variants with further filter flags occur in the <code>SomaticCombineChannel</code> and <code>SomaticAnnotateMaf</code> processes. The union of variants that pass the somatic scoring models intrinsic to the callers (<code>FILTER=&quot;PASS&quot;</code> in the VCF files) are combined, giving precedence to MuTect2 for any site where both callers detected a variant.</p> <p>The functional effect of variants is predicted using <a href="https://www.ensembl.org/vep" target="_blank" rel="noopener noreferrer">VEP<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a> using <a href="https://github.com/mskcc/vcf2maf" target="_blank" rel="noopener noreferrer">vcf2maf<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>, which also converts from VCF into a tab-delimited MAF file. See notes on use of <a href="/reference-resources.html#preferred-transcript-isoforms">preferred transcript isoforms</a> and <a href="https://useast.ensembl.org/info/docs/tools/vep/vep_formats.html#output" target="_blank" rel="noopener noreferrer">VEP annotation outputs<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>.</p> <p>The following columns are added to the final MAF file, in addition to those added during the VEP annotation:</p> <ul><li><code>Strelka2FILTER</code>: Indicates that Strelka2 detected the variant but did not classify it as a somatic variant.</li> <li><code>gnomAD_FILTER</code>: Indicates that the variant was detected in the gnomAD workflow, but ultimately <em>not</em> classified as a <em>germline</em> variant. Note that this is not used in current filtering schema.</li> <li><code>RepeatMasker</code> and <code>EncodeDacMapability</code>: The variant locus is in a repeat, low-mappability, or hard-to-sequence region. More details in the <a href="/reference-resources.html#repeatmasker-and-mappability-blacklist">reference file description</a>.</li> <li><code>PoN</code>: Panel-of-normals, the number of normal samples in which the variant was detected. <a href="/wes-panel-of-normals.html">More details on the implementation for exomes</a>. Under development for genomes, currently applies exome PoN.</li> <li><code>Ref_Tri</code>: Trinucleotide context of SNVs, normalized to pyrimidine-to-purine transversions.</li> <li>gnomAD allele frequencies, <a href="/gnomad.html">more details here</a>:
<ul><li>Exomes, columns named <code>non_cancer_*</code>: Allele counts (<code>AC</code>) and frequencies (<code>AF</code>) for the variant in the non-TCGA population and sub-populations of gnomAD.</li> <li>Genomes, columns named (<code>AC_*</code>|<code>AF_*</code>): Allele counts and frequencies for the variant in populations and sub-populations of gnomAD.</li></ul></li> <li>Raw allele counts: total and strand specific (<code>*_fwd</code> and <code>*_rev</code>) allele <code>count</code> and <code>depth</code> for tumor (<code>t_*</code>) and normal (<code>n_*</code>). These are unfiltered values, in contrast to those from the variant callers, generated by <a href="https://github.com/zengzheng123/GetBaseCountsMultiSample" target="_blank" rel="noopener noreferrer">GetBaseCountsMultiSample<svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg></a>.</li> <li><code>alt_bias</code>: For variants with a <em>raw</em> depth of at least 5 reads, this is true if all <em>raw</em> variant-supporting reads are on either the forward or reverse strand.</li> <li><code>ref_bias</code>: For variants with a <em>raw</em> depth of at least 5 reads, this is true if all <em>raw</em> reads are on either the forward or reverse strand.</li> <li>Mutation hotspots, also see :
<ul><li><code>snv_hotspot</code>: SNV hotspots.</li> <li><code>threeD_hotspot</code>: 3D hotspots, in contrast to abovementioned &quot;linear&quot; hotspots.</li> <li><code>indel_hotspot</code> and <code>indel_hotspot_type</code>: In-frame indel indel hotspots, and indication of whether it overlaps a prior indel hotspot locus (<code>prior</code>) or overlaps an SNV hotspot (<code>novel</code>).</li> <li><code>Hotspot</code>: <code>TRUE</code> if either a linear SNV or indel hotspot.</li></ul></li> <li>OncoKB annotation:
<ul><li><code>mutation_effect</code> and <code>oncogenic</code>: Indicate the functional effect of the mutation and whether it is deemed to be oncogenic</li> <li><code>LEVEL_*</code> and <code>Highest_level</code>: Indicates whether there is any drug at the given level of actionability and which is the highest level of actionability, if any. Note that this is cancer-type agnostic in current implementation.</li> <li><code>citations</code>: References for OncoKB annotation.</li></ul></li> <li>Variant caller metadata (development feature, subject to changes):
<ul><li>MuTect2<sup>1</sup>:
<ul><li><code>MBQ</code>: Median base quality, comma-separated for reference and alternate allele.</li> <li><code>MFRL</code>: Median fragment length, comma-separated for reference and alternate allele.</li> <li><code>MMQ</code>: Median mapping quality, comma-separated for reference and alternate allele.</li> <li><code>MPOS</code>: Median distance of variant from end of read.</li> <li><code>OCM</code>: Number of reads whose original alignment does not match the reference.</li> <li><code>RPA</code>: If tandem repeat, number of times repeated (can be comma-separated for reference and alternate allele).</li> <li><code>STR</code>: Boolean, indicating that variant is a short tandem repeat.</li> <li><code>ECNT</code>: Number of events in haplotype.</li></ul></li> <li>Strelka2<sup>2</sup>:
<ul><li><code>MQ</code>: Root mean square mapping quality.</li> <li><code>SNVSB</code>: Strand bias for somatic SNVs.</li> <li><code>FDP</code>: Number of basecalls filtered from original read depth for tier 1* read counts, for tumor (<code>t_FDP</code>) and normal (<code>n_FDP</code>).</li> <li><code>SUBDP</code>: Number of reads below tier 1 mapping-quality threshold aligned across site, for tumor (<code>t_SUBDP</code>) and normal (<code>n_SUBDP</code>).</li> <li><code>RU</code>: If indel, smallest repeating sequence unit in inserted or deleted sequence.</li> <li><code>IC</code>: If indel, number of times <code>RU</code> is repeated in variant.</li></ul></li></ul></li></ul> <p>The <code>FILTER</code> column in the unfiltered MAF file, can contain any semi colon-separated combination of the following filter flags, or say <code>PASS</code>:</p> <ul><li><code>part_of_mnv</code>: The variant is likely part of another called multi-nucleotide variant (MNV).</li> <li><code>multiallelic2</code>: Multiallelic loci, likely artifact. For variants called by Strelka2. The <code>2</code> is added due the presence of <code>multiallelic</code> flag in the MuTect2 VCFs.</li> <li><code>strand_bias</code>, variants likely artifactual due to strand bias:
<ul><li>For variants called by Mutect2, if all supporting reads come from one strand and there are a least 10 reads on both strands in either normal or tumor sample.</li> <li>For variants called by Strelka2, if the total alternate read count is above 10 and all of these fall on either strand; or low mapping-quality variant suffering from bias in both supporting reads and total reads.</li></ul></li> <li><code>caller_conflict</code>: Variant was detected by both callers, but did not pass Strelka2's thresholds for somatic variant calling.</li> <li>The following read depth-based flags are parameterized according to the sequencing platform, see the <code>exome.config</code> and <code>genome.config</code> files.
<ul><li><code>low_vaf</code>: Variant falls below lower threshold for tumor variant allele fraction (VAF).</li> <li><code>low_t_depth</code>: Variant falls below lower threshold for total depth in the tumor.</li> <li><code>low_t_alt_count</code>: Variant falls below lower threshold for reads supporting variant allele in tumor.</li> <li><code>low_n_depth</code>: Variant falls below lower threshold for total depth in normal.</li> <li><code>high_n_alt_count</code>: Variant exceeds upper threshold for reads supporting variant allele normal.</li> <li><code>mappability</code>/<code>repeatmasker</code>: Variant falls in blacklisted genomic region.</li> <li><code>high_gnomad_pop_af</code>: Variant exceeds upper threshold for allele fraction in gnomAD.</li> <li><code>PoN</code>: Variant exceeds upper threshold for count in panel of normals.</li> <li><code>low_mapping_quality</code>: For indels called by Strelka2, variant falls below lower mapping quality threshold.</li></ul></li></ul> <p><small><sup>1</sup>See the MuTect2 documentation for more information: https://software.broadinstitute.org/gatk/documentation/article?id=11005</small><br> <small><sup>2</sup>See the Strelka2 documentation for more information: https://github.com/Illumina/strelka/blob/v2.9.x/docs/userGuide/README.md</small></p> <h3 id="whitelisting"><a href="#whitelisting" aria-hidden="true" class="header-anchor">#</a> Whitelisting</h3> <p>Mutational hotspots, where the value in <code>Hotspot</code> is <code>TRUE</code>, are retained in the filtered MAF file, if they:</p> <ul><li>Are flagged with <code>low_vaf</code> but the tumor VAF is at least 0.02.</li> <li>Are flagged with <code>low_mapping_quality</code>, <code>low_t_depth</code>, or <code>strand_bias</code>.
<em>Note that combinations of above filter flags results in filtering of the variant.</em></li></ul> <h3 id="clonality-and-zygosity-analyses"><a href="#clonality-and-zygosity-analyses" aria-hidden="true" class="header-anchor">#</a> Clonality and Zygosity Analyses</h3> <h2 id="germline-snvs-and-indels"><a href="#germline-snvs-and-indels" aria-hidden="true" class="header-anchor">#</a> Germline SNVs and Indels</h2> <p>TBD</p> <h3 id="zygosity-analysis"><a href="#zygosity-analysis" aria-hidden="true" class="header-anchor">#</a> Zygosity Analysis</h3> <h2 id="somatic-and-germline-svs"><a href="#somatic-and-germline-svs" aria-hidden="true" class="header-anchor">#</a> Somatic and Germline SVs</h2> <p>TBD</p></div> <footer class="page-edit"><!----> <!----></footer> <div class="page-nav"><p class="inner"><span class="prev">
        ←
        <a href="/reference-resources.html" class="prev">
          Reference resources
        </a></span> <span class="next"><a href="/troubleshooting.html">
          Troubleshooting
        </a>
        →
      </span></p></div> </main></div><div class="global-ui"></div></div>
    <script src="/assets/js/app.2b579abf.js" defer></script><script src="/assets/js/2.e10afc37.js" defer></script><script src="/assets/js/18.fedd8970.js" defer></script>
  </body>
</html>
