(window.webpackJsonp=window.webpackJsonp||[]).push([[19],{202:function(e,t,a){"use strict";a.r(t);var r=a(0),s=Object(r.a)({},function(){var e=this,t=e.$createElement,a=e._self._c||t;return a("ContentSlotsDistributor",{attrs:{"slot-key":e.$parent.slotKey}},[a("h1",{attrs:{id:"reference-resources"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#reference-resources","aria-hidden":"true"}},[e._v("#")]),e._v(" Reference resources")]),e._v(" "),a("h2",{attrs:{id:"repeatmasker-and-mapability-blacklist"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#repeatmasker-and-mapability-blacklist","aria-hidden":"true"}},[e._v("#")]),e._v(" RepeatMasker and mapability blacklist")]),e._v(" "),a("p",[e._v("BED files with with genomic repeat and mapability information are used to annotate the VCFs with somatic and germline SNV/indels. These data are from "),a("a",{attrs:{href:"http://www.repeatmasker.org/",target:"_blank",rel:"noopener noreferrer"}},[e._v("RepeatMasker"),a("OutboundLink")],1),e._v(" and the "),a("a",{attrs:{href:"http://rohsdb.cmb.usc.edu/GBshape/ENCODE/index.html",target:"_blank",rel:"noopener noreferrer"}},[e._v("ENCODE consortium"),a("OutboundLink")],1),e._v(", and the files are retrieved from the "),a("a",{attrs:{href:"https://genome.ucsc.edu",target:"_blank",rel:"noopener noreferrer"}},[e._v("UCSC Genome Browser"),a("OutboundLink")],1),e._v(" and parsed as such:")]),e._v(" "),a("div",{staticClass:"language-shell extra-class"},[a("pre",{pre:!0,attrs:{class:"language-text"}},[a("code",[e._v("wget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/rmsk.txt.gz\ngunzip rmsk.txt.gz\ncut -f6-8,12 rmsk.txt | \\\n    grep -e \"Low_complexity\" -e \"Simple_repeat\" | \\\n    sed 's/^chr//g'> rmsk_mod.bed\nbgzip rmsk_mod.bed\ntabix --preset bed rmsk_mod.bed.gz\n\nwget http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeMapability/wgEncodeDacMapabilityConsensusExcludable.bed.gz\ngunzip wgEncodeDacMapabilityConsensusExcludable.bed.gz\nsed -i 's/^chr//g' wgEncodeDacMapabilityConsensusExcludable.bed\nbgzip wgEncodeDacMapabilityConsensusExcludable.bed\ntabix --preset bed wgEncodeDacMapabilityConsensusExcludable.bed.gz\n")])])]),a("h2",{attrs:{id:"preferred-transcript-isoforms"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#preferred-transcript-isoforms","aria-hidden":"true"}},[e._v("#")]),e._v(" Preferred transcript isoforms")]),e._v(" "),a("p",[e._v("The "),a("code",[e._v("--custom-enst")]),e._v(" argument to vcf2maf takes a list of preferred gene transcript isoforms which mtations are mapped onto. We supply a consensus list of "),a("a",{attrs:{href:"https://github.com/mskcc/vcf2maf/tree/master/data",target:"_blank",rel:"noopener noreferrer"}},[a("code",[e._v("isoform_overrides_at_mskcc")]),e._v(" and "),a("code",[e._v("isoform_overrides_uniprot")]),a("OutboundLink")],1),e._v(", generated as such:")]),e._v(" "),a("div",{staticClass:"language-r extra-class"},[a("pre",{pre:!0,attrs:{class:"language-r"}},[a("code",[e._v("t1 "),a("span",{pre:!0,attrs:{class:"token operator"}},[e._v("=")]),e._v(" readr"),a("span",{pre:!0,attrs:{class:"token operator"}},[e._v("::")]),e._v("read_tsv"),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v("(")]),a("span",{pre:!0,attrs:{class:"token string"}},[e._v("'isoform_overrides_at_mskcc'")]),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v(")")]),e._v("\nt2 "),a("span",{pre:!0,attrs:{class:"token operator"}},[e._v("=")]),e._v(" readr"),a("span",{pre:!0,attrs:{class:"token operator"}},[e._v("::")]),e._v("read_tsv"),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v("(")]),a("span",{pre:!0,attrs:{class:"token string"}},[e._v("'isoform_overrides_uniprot'")]),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v(")")]),e._v("\nt2 "),a("span",{pre:!0,attrs:{class:"token percent-operator operator"}},[e._v("%>%")]),e._v("\n    dplyr"),a("span",{pre:!0,attrs:{class:"token operator"}},[e._v("::")]),e._v("filter"),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v("(")]),e._v("gene_name "),a("span",{pre:!0,attrs:{class:"token percent-operator operator"}},[e._v("%nin%")]),e._v(" t1"),a("span",{pre:!0,attrs:{class:"token operator"}},[e._v("$")]),e._v("gene_name"),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v(")")]),e._v(" "),a("span",{pre:!0,attrs:{class:"token percent-operator operator"}},[e._v("%>%")]),e._v("\n    dplyr"),a("span",{pre:!0,attrs:{class:"token operator"}},[e._v("::")]),e._v("bind_rows"),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v("(")]),e._v("."),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v(",")]),e._v(" t1"),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v(")")]),e._v(" "),a("span",{pre:!0,attrs:{class:"token percent-operator operator"}},[e._v("%>%")]),e._v("\n    readr"),a("span",{pre:!0,attrs:{class:"token operator"}},[e._v("::")]),e._v("write_tsv"),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v("(")]),a("span",{pre:!0,attrs:{class:"token string"}},[e._v("'isoforms'")]),a("span",{pre:!0,attrs:{class:"token punctuation"}},[e._v(")")]),e._v("\n")])])]),a("h2",{attrs:{id:"hotspot-annotation"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#hotspot-annotation","aria-hidden":"true"}},[e._v("#")]),e._v(" Hotspot annotation")]),e._v(" "),a("p",[e._v("Three types of mutation hotspots are annotated in the somatic MAF. These include SNV, indel in linear space as well as SNV hotspots in 3D space. These are annotated with the "),a("a",{attrs:{href:"https://github.com/taylor-lab/annotateMaf",target:"_blank",rel:"noopener noreferrer"}},[e._v("annotateMaf package"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("h2",{attrs:{id:"oncokb"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#oncokb","aria-hidden":"true"}},[e._v("#")]),e._v(" OncoKB")]),e._v(" "),a("p",[e._v("Functional mutation effects and predicted oncogenicity of variants, as well as level of clinical actionability are from "),a("a",{attrs:{href:"https://oncokb.org",target:"_blank",rel:"noopener noreferrer"}},[e._v("OncoKB"),a("OutboundLink")],1),e._v(" and annotated using the "),a("a",{attrs:{href:"https://github.com/oncokb/oncokb-annotator",target:"_blank",rel:"noopener noreferrer"}},[e._v("OncoKB annotator"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("h2",{attrs:{id:"grch37"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#grch37","aria-hidden":"true"}},[e._v("#")]),e._v(" GRCh37")]),e._v(" "),a("h3",{attrs:{id:"genome"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#genome","aria-hidden":"true"}},[e._v("#")]),e._v(" Genome")]),e._v(" "),a("p",[a("a",{attrs:{href:"https://software.broadinstitute.org/gatk/download/bundle",target:"_blank",rel:"noopener noreferrer"}},[e._v("GATK bundle"),a("OutboundLink")],1),e._v(", also available "),a("a",{attrs:{href:"https://console.cloud.google.com/storage/browser/gatk-legacy-bundles/b37",target:"_blank",rel:"noopener noreferrer"}},[e._v("here"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("h3",{attrs:{id:"snv-and-indel-calling"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#snv-and-indel-calling","aria-hidden":"true"}},[e._v("#")]),e._v(" SNV and InDel calling")]),e._v(" "),a("p",[e._v("For genomes, use "),a("a",{attrs:{href:"https://github.mskcc/vaporware/blob/master/docs/INTERVALS.md#genome",target:"_blank",rel:"noopener noreferrer"}},[a("code",[e._v("bed")]),e._v(' file of "callable"'),a("OutboundLink")],1),e._v(" regions from GATK's bundle.")]),e._v(" "),a("p",[e._v("For exomes, use "),a("code",[e._v("bed")]),e._v(" file corresponding to the platform used for target capture, see "),a("a",{attrs:{href:"https://github.mskcc/vaporware/blob/master/docs/INTERVALS.md#exome-capture-platform",target:"_blank",rel:"noopener noreferrer"}},[e._v("documentation on intervals"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("h3",{attrs:{id:"structural-variant-sv-calling"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#structural-variant-sv-calling","aria-hidden":"true"}},[e._v("#")]),e._v(" Structural Variant (SV) calling")]),e._v(" "),a("p",[e._v("Delly provides and takes as an argument a "),a("a",{attrs:{href:"https://github.com/dellytools/delly/tree/master/excludeTemplates",target:"_blank",rel:"noopener noreferrer"}},[e._v("file of regions"),a("OutboundLink")],1),e._v(" to "),a("em",[e._v("exclude")]),e._v(" from variant calling. This excludes telomeres and centromeres from auto- and allosomes as well as any other contig.")]),e._v(" "),a("p",[e._v("For Manta, subtract these regions from a bed file of the whole genome to generate a list of regions to "),a("em",[e._v("include")]),e._v(". First clean up the file provided by Delly, since it is not in "),a("code",[e._v("bed")]),e._v(" format:")]),e._v(" "),a("div",{staticClass:"language-shell extra-class"},[a("pre",{pre:!0,attrs:{class:"language-text"}},[a("code",[e._v('grep -Ev "chr|MT|GL00|NC|hs37d5" human.hg19.excl.tsv > human.hg19.excl.clean.bed\nbedtools subtract -a b37.bed -b human.hg19.excl.clean.bed > b37.minusDellyExclude.bed\n')])])]),a("h2",{attrs:{id:"grch38"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#grch38","aria-hidden":"true"}},[e._v("#")]),e._v(" GRCh38")]),e._v(" "),a("p",[a("a",{attrs:{href:"https://software.broadinstitute.org/gatk/download/bundle",target:"_blank",rel:"noopener noreferrer"}},[e._v("GATK bundle"),a("OutboundLink")],1),e._v(", also available "),a("a",{attrs:{href:"https://console.cloud.google.com/storage/browser/genomics-public-data/resources/broad/hg38",target:"_blank",rel:"noopener noreferrer"}},[e._v("here"),a("OutboundLink")],1),e._v(".")]),e._v(" "),a("p",[a("strong",[e._v("Note:")]),e._v(" Support for hg38 is currently somewhat limited. Please raise an issue at https://github.com/mskcc/vaporware/issues if you would like to process data with GRch38. However, please note that hg38 reference files are easily available from UCSC.")])])},[],!1,null,null,null);t.default=s.exports}}]);